extensions/amp-truncate-text/1.0/truncation.js

```diff
@@ -0,0 +1,345 @@
+ * limitations under the License.
+ */
+
+import {
+  BinarySearchPreference,
+  BinarySearchStop,
+  binarySearch,
+} from './binary-search';
+import {trimEnd} from '../../../src/string';
+
+/** The class to add to the container when it has overflow. */
+const CONTAINER_OVERFLOW_ATTRIBUTE = 'i-amphtml-truncate-overflow';
+
+/** The class to add to a descendant element that is overflowing. */
+const ELEMENT_OVERFLOW_ATTRIBUTE = 'i-amphtml-truncate-child-overflow';
+
+/** Used to save the Text Node's data original and modified data.  */
+const TEXT_DATA_PROPERTY = '__AMP_TRUNCATE_TEXT_DATA';
+
+/** A marker property for when an element has been truncateed. */
+const TRUNCATED_MARKER = '__AMP_TRUNCATE_TRUNCATED';
+
+/**
+ * The characters to use when ellipsizing. This includes a space to add some
+ * space between the ellipsis and the overflow element. Note that if there is
+ * no overflow element, the space character will not take any extra space.
+ */
+const ELLIPSIS_CHARACTERS = '… ';
+
+/**
+ * @param {string} char
+ * @return {boolean} True if the character is whitespace, false otherwise.
+ */
+function isWhitespace(char) {
+  return !char.trim();
+}
+
+/**
+ * Clears the effects of truncateing on an element, restoring content, unhiding
+ * any hidden elements and hiding the overflow button.
+ * @param {!Element} element
+ */
+function clearTruncated(element) {
+  // Save some time on the first call, since we do not need to clear anything
+  // yet,
+  if (!element[TRUNCATED_MARKER]) {
+    return;
+  }
+
+  element[TRUNCATED_MARKER] = false;
+  element.removeAttribute(CONTAINER_OVERFLOW_ATTRIBUTE);
+  removeTruncation(element);
+}
+
+/**
+ * Marks an element as truncateed, showing the overflow button.
+ * @param {!Element} element
+ */
+function setTruncated(element) {
+  element[TRUNCATED_MARKER] = true;
+  element.setAttribute(CONTAINER_OVERFLOW_ATTRIBUTE, '');
+}
+
+/**
+ * @param {!Node} node The node to operate on.
+ * @param {function(!Node)} cb  A callback to call for each child.
+ */
+function forEachChild(node, cb) {
+  // Check for slots in the light DOM, which may have content projected into
+  // them. For example:
+  //
+  // <amp-slide>
+  //   # ShadowDOM
+  //   <amp-truncate-text>
+  //     <slot name="caption"></slot>
+  //     <button slot="expand">More</button>
+  //   </amp-truncate-text>
+  // </amp-slide>
+  //
+  // <amp-slide>
+  //   <div slot="caption">...</div>
+  // </amp-slide>
+  //
+  // We want to get the contents of the "caption" slot and iterate over them.
+  // We use flatten so that we get the slot default contents, if nothing is
+  // slotted.
+  //
+  // In Preact mode, a slot may exist in the light DOM, so we fallback to
+  // `childNodes` if `assignedNodes` comes up empty.
+  let {childNodes} = node;
+  if (node.localName == 'slot') {
+    const assignedNodes = node.assignedNodes({flatten: true});
+    if (assignedNodes.length) {
+      childNodes = assignedNodes;
+    }
```

Comment:
This section was previously
```javascript
  const childNodes =
    node.localName == 'slot'
      ? node.assignedNodes({flatten: true})
      : node.childNodes;
```

But in Preact mode, `assignedNodes` comes up empty for a `<slot />`. IIUC, I shouldn't have an explicit slot in Preact mode, so this change may go away. TBD